pkg/dyninst/decode/data_item_iterator.go

```diff
@@ -0,0 +1,178 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package decode
+
+import (
+	"errors"
+	"fmt"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/dyninst/output"
+)
+
+const (
+	eventHeaderSize    = int(unsafe.Sizeof(output.EventHeader{}))
+	dataItemHeaderSize = int(unsafe.Sizeof(output.DataItemHeader{}))
+)
+
+var finishedIterating = errors.New("no data items left to iterate")
+
+type dataItem struct {
+	header *output.DataItemHeader
+	data   []byte
+}
+
+type dataEventIterator struct {
+	buf            []byte
+	eventHeader    *output.EventHeader
+	idx            int
+	addressPassMap map[uint64]*dataItem
+}
```

Comment:
This thing is doing more than being an iterator. In particular, it seems odd to me for it to have anything to do with holding onto this map. The user of this thing should be interacting with the map. 

Here's what I think I would do for this:

```go
type dataItem struct {
    header *output.DataItemHeader
    data []byte
}

type Event []byte

func (e Event) eventHeader() (*output.EventHeader, error) { ... }

func (e Event) dataItems() iter.Seq2[dataItem, error] { ... }

// An optimization:
func (e Event) firstDataItem() (dataItem, error) { ... } 

func (e Event) stackFrames() ([]uint64, error) { ... }
```

In this model I think we avoid all allocations.